### PR TITLE
nnbd for integration_test

### PR DESCRIPTION
**Pull request by dnfield**
TBD: What version/changelog changes to make.

### CODE REVIEW

**Comment by blasten** [comment]
`1.0.0-nullsafety`?

**Comment by dnfield** [comment]
Are we going to use teh `-nullsafety` for this? I was under the impression that that is a temporary designation for special packages and should not be widely used.

**Comment by blasten** [comment]
Yes, the `-nullsafety` prerelease is temporary. The general guidance is to consider this as a breaking change (major version bump), so you can clean up/refine the API as needed.

File: packages/integration_test/example/pubspec.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ description: Demonstrates how to use the integration_test plugin.
 publish_to: 'none'
 
 environment:
-  sdk: ">=2.1.0 <3.0.0"
+  sdk: '>=2.10.0-56.0.dev <3.0.0'
   flutter: ">=1.6.7 <2.0.0"
```
Comment (CareF) [inline packages/integration_test/example/pubspec.yaml:7]: Should we also update flutter version?

File: packages/integration_test/lib/_callback_web.dart [hunk 2]
```diff
@@ -75,14 +75,13 @@ class WebCallbackManager implements CallbackManager {
...
       Map<String, String> params, IntegrationTestResults testRunner) async {
-    final String command = params['command'];
+    final String command = params['command'] as String;
     Map<String, String> response;
     switch (command) {
       case 'request_data':
         return params['message'] == null
             ? _requestData(testRunner)
-            : _requestDataWithMessage(params['message'], testRunner);
-        break;
```
Comment (CareF) [inline packages/integration_test/lib/_callback_web.dart:85]: Why remove this `break`?

File: packages/integration_test/lib/integration_test.dart [hunk 3]
```diff
@@ -195,28 +195,29 @@ class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
...
 
   /// Initialize the [vm.VmService] settings for the timeline.
   @visibleForTesting
   Future<void> enableTimeline({
     List<String> streams = const <String>['all'],
-    @visibleForTesting vm.VmService vmService,
+    @visibleForTesting vm.VmService? vmService,
   }) async {
-    assert(streams != null);
+    assert(streams != null); // ignore: unnecessary_null_comparison
```
Comment (CareF) [inline packages/integration_test/lib/integration_test.dart:217]: This line seems no longer necessary when we have null-safety.

File: packages/integration_test/lib/integration_test.dart [hunk 4]
```diff
@@ -195,28 +195,29 @@ class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
...
-    if (_vmService == null) {
-      final developer.ServiceProtocolInfo info =
-          await developer.Service.getInfo();
-      assert(info.serverUri != null);
-      _vmService = await vm_io.vmServiceConnectUri(
-        'ws://localhost:${info.serverUri.port}${info.serverUri.path}ws',
-      );
-    }
-    await _vmService.setVMTimelineFlags(streams);
+    vmService ??= await _vmService;
```
Comment (CareF) [inline packages/integration_test/lib/integration_test.dart:219]: I think this should be `if (vmService != null) _cachedVmService = vmService`, so that `traceTimeline` will use the provided `vmService`.

File: packages/integration_test/lib/integration_test_driver.dart [hunk 5]
```diff
@@ -30,9 +30,9 @@ typedef ResponseDataCallback = FutureOr<void> Function(Map<String, dynamic>);
 Future<void> writeResponseData(
   Map<String, dynamic> data, {
   String testOutputFilename = 'integration_response_data',
-  String destinationDirectory,
+  String? destinationDirectory,
 }) async {
-  assert(testOutputFilename != null);
+  assert(testOutputFilename != null); // ignore: unnecessary_null_comparison
```
Comment (CareF) [inline packages/integration_test/lib/integration_test_driver.dart:35]: This line seems no longer necessary when we have null-safety.

File: packages/integration_test/lib/integration_test.dart (hunk 3 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart:217]: It's still useful when not in sound null safety e.g. importing a non nnbd package

File: packages/integration_test/lib/integration_test.dart (hunk 4 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart:219]: I've made the private member a future. I'm also not clear on this, I assumed the param was only for testing.

File: packages/integration_test/lib/_callback_web.dart (hunk 2 above)
Comment (dnfield) [reply packages/integration_test/lib/_callback_web.dart:85]: The line before it is a return.  It's dead code

File: packages/integration_test/lib/integration_test.dart (hunk 4 above)
Comment (CareF) [reply packages/integration_test/lib/integration_test.dart:219]: But this parameter is intended to replace the standard `vmService` for testing purposes. If I understand it correctly, your version `traceTimeline` will still call the `_cachedVmService` obtained by `vm_io.vmServiceConnectUri`, but what it should call is the `vmService` provided here.

Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart:219]: Hmm.  We should add it as a param there then or move it to the ctor/ensureInitialized  It's pretty strange to have a method alter a private member for tests.

File: packages/integration_test/example/pubspec.yaml (hunk 1 above)
Comment (dnfield) [reply packages/integration_test/example/pubspec.yaml:7]: I think the Dart constraint will take care of that, but we could.

File: packages/integration_test/lib/integration_test_driver.dart (hunk 5 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test_driver.dart:35]: Same comment as above - these are still helpful when working in mixed/unsound mode.

File: packages/integration_test/lib/integration_test.dart (hunk 4 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart]: I've cleaned this up a bit, let me know what you think.

**Comment by dnfield** [comment]
This is going to have problems because of package:mockito and package:vm_service.

**Comment by dnfield** [comment]
So we don't really need mockito for this, and the story around using that for NNBD is still being developed so we can just avoid having it block this.

Still will be blocked on migration of package:vm_service

File: packages/integration_test/lib/integration_test.dart [hunk 6]
```diff
@@ -200,28 +205,28 @@ class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
     results[description] ??= _success;
   }
 
-  vm.VmService _vmService;
+  vm.VmService? _cachedVmService;
+  Future<vm.VmService> get _vmService async {
```
Comment (nturgut) [inline packages/integration_test/lib/integration_test.dart:209]: Is this change related to nnbd?

File: packages/integration_test/pubspec.yaml [hunk 7]
```diff
@@ -1,10 +1,10 @@
 name: integration_test
 description: Runs tests that use the flutter_test API as integration tests.
-version: 0.9.2+1
+version: 0.10.0-nullsafety
 homepage: https://github.com/flutter/plugins/tree/master/packages/integration_test
 
 environment:
-  sdk: ">=2.2.2 <3.0.0"
+  sdk: '>=2.10.0-56.0.dev <3.0.0'
```
Comment (nturgut) [inline packages/integration_test/pubspec.yaml:7]: I wonder if this will break the flutter/engine tests, when we want to update e2e version?

https://github.com/flutter/engine/blob/master/e2etests/web/regular_integration_tests/pubspec.yaml#L5 For integration tests

web_ui: https://github.com/flutter/engine/blob/master/lib/web_ui/pubspec.yaml#L5

**Review by nturgut — COMMENTED** [review]
I added a few comments. I think the best way would be to test a web engine integration test locally with the new version (since the existing tests use the old version we won't know if they fail until we try to upgrade)

File: packages/integration_test/lib/integration_test.dart (hunk 6 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart:209]: This could be split out into a separate patch. It came up because with nnbd I had to think more carefully about whether this property was nullable or not, and when it'd be set, and came to realize it was getting set by a method when another method might want to use it too.

File: packages/integration_test/pubspec.yaml (hunk 7 above)
Comment (dnfield) [reply packages/integration_test/pubspec.yaml:7]: There's no rush on this, and we won't be able to move forward with it realistically until package:vm_service is migrated, which probably won't happen for at least a couple more weeks.

This constraint _should_ be ok, but by the time we'd publish this it should have a real 2.11 (not .dev) constraint.

Comment (nturgut) [reply packages/integration_test/pubspec.yaml:7]: thanks for info! :)

File: packages/integration_test/lib/integration_test.dart (hunk 6 above)
Comment (nturgut) [reply packages/integration_test/lib/integration_test.dart:209]: No worries, let's keep it here. I think it's enough as long as we add it to the change log/pr description (if something goes wrong it will be easier to see 😅 )

File: packages/integration_test/lib/integration_test.dart (hunk 3 above)
Comment (nturgut) [reply packages/integration_test/lib/integration_test.dart:226]: nit: does it make sense to merge to asserts?

**Review by nturgut — APPROVED** [review]

File: packages/integration_test/lib/integration_test.dart (hunk 3 above)
Comment (dnfield) [reply packages/integration_test/lib/integration_test.dart:226]: @nturgut I'm not sure I'm understanding what merge asserts would be in this context...

Comment (nturgut) [reply packages/integration_test/lib/integration_test.dart:226]: it's very nit :) 

assert(streams !=null && streams. isNotEmpty, 'Non empty stream is required')

**Comment by dnfield** [comment]
I'm going to close this, since this package is moving to the flutter repo.

**Comment by ferhatb** [comment]
@dnfield this is blocking us from migrating our tests to non-nullable. Is there a separate issue that tracks move to flutter repo ?

**Comment by dnfield** [comment]
It's been moved already. The real blocker for this is that vm_service and then flutter_driver need to be migrated
